test/e2e/init_scale_zero_test.go

```diff
@@ -0,0 +1,313 @@
+		newCM.Data["check-validity-on-deploy"] = "true"
+		test.CleanupOnInterrupt(func() { restoreClusterDefaults(t, clients, newCM) })
+		defer restoreClusterDefaults(t, clients, newCM)
+	}
+
+	t.Log("Creating a new Service without check validity on deploy being set.")
+	objects, _, err := v1a1test.CreateRunLatestServiceReady(t, clients, &namesCheckValidityUnset,
+		false, /* https TODO(taragu) turn this on after helloworld test running with https */)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", namesCheckValidityUnset.Service, err)
+	}
+	// Verify that no pods are created
+	pods := clients.KubeClient.Kube.CoreV1().Pods(test.ServingNamespace)
+	podList, err := pods.List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", serving.RevisionLabelKey, objects.Revision.Name),
+	})
+	if err != nil {
+		t.Fatalf("Failed to list all pods: %v: %v", namesCheckValidityUnset.Service, err)
+	}
+	if len(podList.Items) != 0 {
+		t.Fatalf("Did not scale to 0; %d pods are created for revision %s.", len(podList.Items), objects.Revision.Name)
+	}
+
+	t.Log("Creating a new Service with check validity on deploy being false.")
+	objects, _, err = v1a1test.CreateRunLatestServiceReady(t, clients, &namesCheckValidityFalse,
+		false, /* https TODO(taragu) turn this on after helloworld test running with https */
+		v1a1testing.WithConfigAnnotations(map[string]string{
+			autoscaling.CheckValidityOnDeployAnnotation: "false",
+		}))
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", namesCheckValidityFalse.Service, err)
+	}
+
+	// Verify that no pods are created
+	pods = clients.KubeClient.Kube.CoreV1().Pods(test.ServingNamespace)
+	podList, err = pods.List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", serving.RevisionLabelKey, objects.Revision.Name),
+	})
+	if err != nil {
+		t.Fatalf("Failed to list all pods: %v: %v", namesCheckValidityFalse.Service, err)
+	}
+	if len(podList.Items) != 0 {
+		t.Fatalf("Did not scale to 0; %d pods are created.", len(podList.Items))
+	}
+
+	t.Log("Creating a new Service with check validity on deploy being true.")
+	objects, _, err = v1a1test.CreateRunLatestServiceReady(t, clients, &namesCheckValidityTrue,
+		false, /* https TODO(taragu) turn this on after helloworld test running with https */
+		v1a1testing.WithConfigAnnotations(map[string]string{
+			autoscaling.CheckValidityOnDeployAnnotation: "true",
+		}))
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", namesCheckValidityTrue.Service, err)
+	}
+
+	// Verify that pods are created
+	pods = clients.KubeClient.Kube.CoreV1().Pods(test.ServingNamespace)
+	podList, err = pods.List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", serving.RevisionLabelKey, objects.Revision.Name),
+	})
+	if err != nil {
+		t.Fatalf("Failed to list all pods: %v: %v", namesCheckValidityTrue.Service, err)
+	}
+	if len(podList.Items) == 0 {
+		t.Fatal("No pods are created.")
+	}
```

Comment:
looks like the behavior is more or less the same. can we extract the validation and make it so that it reuses the same code for listing pods and doing the validations on the results? similar to patterns we have seen in other tests for building up test scenarios?